src/NSFW.cpp

```diff
@@ -1,311 +1,280 @@
-
-  SetPrototypeMethod(tpl, "start", Start);
-  SetPrototypeMethod(tpl, "stop", Stop);
-
-  v8::Local<v8::Context> context = Nan::GetCurrentContext();
-  constructor.Reset(tpl->GetFunction(context).ToLocalChecked());
-  Set(target, New<v8::String>("NSFW").ToLocalChecked(), tpl->GetFunction(context).ToLocalChecked());
-}
-
-NAN_METHOD(NSFW::JSNew) {
-  if (!info.IsConstructCall()) {
-    const int argc = 4;
-    v8::Isolate *isolate = info.GetIsolate();
-    v8::Local<v8::Value> argv[argc] = {info[0], info[1], info[2], info[3]};
-    v8::Local<v8::Context> context = isolate->GetCurrentContext();
-    v8::Local<v8::Function> cons = v8::Local<v8::Function>::New(isolate, constructor);
-    info.GetReturnValue().Set(cons->NewInstance(context, argc, argv).ToLocalChecked());
-    return;
-  }
-
-  if (info.Length() < 1 || !info[0]->IsUint32()) {
-    return ThrowError("First argument of constructor must be a positive integer.");
-  }
-  if (info.Length() < 2 || !info[1]->IsString()) {
-    return ThrowError("Second argument of constructor must be a path.");
-  }
-  if (info.Length() < 3 || !info[2]->IsFunction()) {
-    return ThrowError("Third argument of constructor must be a callback.");
-  }
-  if (info.Length() < 4 || !info[3]->IsFunction()) {
-    return ThrowError("Fourth argument of constructor must be a callback.");
+NSFW::~NSFW() {
+  mErrorCallback.Release();
+  mEventCallback.Release();
+  if (gcEnabled) {
+    instanceCount--;
   }
-
-  v8::Local<v8::Context> context = Nan::GetCurrentContext();
-  uint32_t debounceMS = info[0]->Uint32Value(context).FromJust();
-  Nan::Utf8String utf8Value(Nan::To<v8::String>(info[1]).ToLocalChecked());
-  std::string path = std::string(*utf8Value);
-  Callback *eventCallback = new Callback(info[2].As<v8::Function>());
-  Callback *errorCallback = new Callback(info[3].As<v8::Function>());
-
-  NSFW *nsfw = new NSFW(debounceMS, path, eventCallback, errorCallback);
-  nsfw->Wrap(info.This());
-  info.GetReturnValue().Set(info.This());
 }
 
-NAN_METHOD(NSFW::Start) {
-  Nan::HandleScope scope;
-
-  NSFW *nsfw = ObjectWrap::Unwrap<NSFW>(info.This());
-  if (!nsfw->mInterfaceLockValid) {
-    return ThrowError("NSFW failed to initialize properly. Try creating a new NSFW.");
-  }
-
-  if (
-    info.Length() < 1 ||
-    !info[0]->IsFunction()
-  ) {
-    return ThrowError("Must provide callback to start.");
-  }
-
-  Callback *callback = new Callback(info[0].As<v8::Function>());
-
-  if (nsfw->mInterface != NULL) {
-    v8::Local<v8::Value> argv[1] = {
-      Nan::Error("This NSFW cannot be started, because it is already running.")
-    };
-    callback->Call(1, argv);
-    delete callback;
-    return;
+NSFW::StartWorker::StartWorker(Napi::Env env, NSFW *nsfw):
+  Napi::AsyncWorker(env, "nsfw"),
+  mDeferred(Napi::Promise::Deferred::New(env)),
+  mDidStartWatching(false),
+  mNSFW(nsfw),
+  mShouldUnref(false)
+{}
+
+Napi::Promise NSFW::StartWorker::RunJob() {
+  if (mNSFW->mInterface) {
```

Comment:
It's not actually a race condition though, a value will either be there or it won't. If it was being destroyed while we were calling start, the operation won't start. We don't need to lock on this pointer, because we're not leveraging anything about the pointer except did it exist when we could read it at this point in time. I really don't put much emphasis on correctly queuing a new start operation when a stop operation is happening at this level.

Reply:
The correct way for a user of the library to queue a start after a stop is to wait for the stop promise to complete before running start.

Reply:
Right, but the very act of accessing a variable (in this case checking if the pointer is null) that's modified in another thread is undefined behavior unless they're atomic or synchronized. There's no guarantee that changes made in one thread will be visible in another. So it could return true, it could return false, it could never return true because maybe it cached the initial value in each thread (or put it in a register) and that thread never busted the cache, etc.

That said, if it's incorrect for users to call start while an instance is already running, I don't mind too much.

Reply:
> it could never return true because maybe it cached the initial value in each thread (or put it in a register) and that thread never busted the cache, etc.

If this were ever true, why would leveraging the mutex change this behavior at all?

Reply:
I'm sorry I really don't see how this is a race condition that has any bearing on the correctness of the application. :( I specifically left out the lock here, because it will always be safe to perform this check, and will basically be correct in all cases.

Reply:
Neat. Ian shared https://devblogs.microsoft.com/oldnewthing/20180924-00/?p=99805. Under O3 optimization in G++, this can in fact force-cache the variable in the register on the thread.